text/0049-parse-multiply-by-constant.md

```diff
@@ -0,0 +1,60 @@
+
+## Motivation
+
+Let's first revisit the status quo: The parser separates the operands of a CST representing a (possibly chained) multiplication operation into a non-integer-constant and a list of integer constants, preserving the order of the constants, and then *left-fold* the constant list with the non-constant. For instance, `2*1*content.value` is parsed into `(mul (mul content.value 2) 1)`.
+
+Note that the parser changes the order of operands, which does not alter the evaluation result since Cedar's multiplication operation is associative. However, changing the evaluation order can make error message confusing: For instance, evaluating `2*6*context.value` when `context.value` is `i64::MAX/10` produces an error message like ```integer overflow while attempting to multiply `1844674407370955160` by `6```, contradicting the intuitive left-to-right evaluation order.
+
+The parser's rule to determine if multiplication operations are valid is also confusing. For instance, `2*1*context.value` is valid whereas its variant `(2*1)*context.value` is not. The inconsistency further causes ambiguity in EST to AST conversion: ESTs of the aforementioned example are the same but only the former can be parsed to an AST.
+
+## Detailed design
+
+To resolve the issues, this PR proposes a design to explicitly apply Cedar's left-associativity rule on the CST level and determine validity afterwards. The CSTs, if deemed valid, should be converted to ASTs that have the same order. Note that there will be only two operands for the CST representation of a multiplication operation, as opposed to a list of operands per status quo.
+
+The validity checking algorithm is simple: Count the non-constant-integer operands of a CST representing a multiplication operation recursively and error if it's greater than 1. An example algorithm is as follows.
+
+```Rust
+fn count_non_constant_operands(lhs, rhs) -> bool {
+    match (lhs, rhs) {
+        (cst::Mul(ll, lr), cst::Mul(rl, rr)) => count_non_constant_operands(ll, lr) + count_non_constant_operands(rl, rr),
+        (cst::Mul(ll, lr), Lit(Int(_))) => count_non_constant_operands(ll, lr),
+        (cst::Mul(ll, lr), _) => count_non_constant_operands(ll, lr) + 1,
+        (Lit(Int(_)), cst::Mul(rl, rr)) => count_non_constant_operands(rl, rr),
+        (_, cst::Mul(rl, rr)) => count_non_constant_operands(rl, rr) + 1,
+        (Lit(Int(_)), Lit(Int(_))) => 0,
+        (Lit(Int(_)), _) => 1,
+        (_, Lit(Int(_))) => 1,
+        (_, _) => 2,
+    }
+}
+```
+
+This recursive algorithm validates expressions derived from arbitrary operand associations of a chained multiplication operation. For instance, expressions `1*2*context*3`, `1*(2*context*3)`, and `1*(2*context)*3` are all accepted.
+
+## Drawbacks
+
+The major drawback of this approach is that it induces a breaking change.
+
+## Alternatives
+
+An alternative is to let the ESTs representing multiplication operations to a list of operands, just like what CSTs do.
```

Comment:
Isn’t the alternative also a breaking change?  Plus, the alternative could also make the EST API harder to use.  For example, which way are the clients of the EST going to associate the list of multiplication elements?

Are there any fixes to #619 that don’t introduce breaking changes?  For example, is it possible to change the CST -> EST converter to produce an EST that associates multiplication in exactly the same way as the CST -> AST converter?  That way, CST -> EST -> AST conversion would fail only if CST -> AST conversion would for the same CST.

Reply:
We could likely do this as an extension to the EST grammar without a breaking change. I do agree that having both options for multiplication makes the EST API more complicated and somewhat harder to used.

My understanding is that changing CST->EST alone wouldn't be enough. The expression `(1 * 2)` in an EST is treated as non-constant, so we would also need to update the EST to AST step (but in an obvious way).

Reply:
To expand John's comment: `(1*2)*true`'s CST to AST conversion fails whereas `1*2*true`'s doesn't. But both of them have the same EST because we do not distinguish `(1*2)` and `1*2` on the EST level. If we want to ensure that if CST->AST fails, then CST->EST->AST also does, we may need to check validity (i.e., if there's more than one non-constant) at CST to EST conversion and also operator association. We still need to also check validity during EST to AST conversion because otherwise the users could create something like `principal.value * context.value`. And we want the two checks consistent.